obtainDirectly, validateAccessToken supports https realm urls

rockxwre opened the pull request:
The `GrantManager`'s `obtainDirectly` and `validateAccessToken` functions were always using the http protocol when making Keycloak requests. Since our Keycloak server is only accepting https requests (and `auth-server-url` in keycloak.json is configured as such) this will not work. I think these functions should also support https.
To reduce some duplicate code I created a function `getProtocol` that will return the protocol instance based on an `options` argument.

abstractj commented:
@rockxwre first thank you for these changes. Could you please add an spec [here](https://github.com/keycloak/keycloak-nodejs-auth-utils/blob/master/spec/grant-manager-spec.js) to reproduce the scenario you mentioned?

rockxwre commented:
@abstractj , I'm not sure if extra specs are needed. By changing _keycloak.json_'s `auth-server-url` into a https url, the specs will use https (well, there is a bit more to it when it's having a self-signed certificate).
Maybe all current specs should run on both http as https? In that case you would need two _keycloak.json_ files, one for http and one for https.

abstractj commented:
@helio-frota @lholmquist @lance when you get the chance, could you please take a look?

lholmquist commented:
commit history seems a little messy here.  should of been just a rebase i think

lholmquist commented on lib/grant-manager.js:
```diff
@@ -455,4 +440,15 @@ GrantManager.prototype.getAccount = function(token, callback) {
+    protocol = http;
+  }
+  return protocol;
+}
```
needs a semicolon at the end of this block

lholmquist commented on lib/grant-manager.js:
```diff
@@ -455,4 +440,15 @@ GrantManager.prototype.getAccount = function(token, callback) {
+GrantManager.prototype.getProtocol = function(options) {
+  var protocol = http;
+
+  if ( options.protocol == 'https:' ) {
```
should be "===" instead of "=="

lholmquist commented on lib/grant-manager.js:
```diff
@@ -455,4 +440,15 @@ GrantManager.prototype.getAccount = function(token, callback) {
 };
 
+GrantManager.prototype.getProtocol = function(options) {
+  var protocol = http;
```
if you are setting the protocol value here to http, then perhaps you don't need the else.  you could probably also do this whole function in one line too

rockxwre replied:
I agree. To avoid confusion I thought it was better to leave the existing code unchanged as much as possible. Anyway, I changed it using a ternary operator.

abstractj commented:
@rockxwre I don't think they are extra specs, just test case scenarios to reproduce the changes you did. I'd appreciated if you could do that and update this PR based on @lholmquist's comments.

rockxwre commented:
@abstractj Updated the code as suggested by @lholmquist.
Added a _keycloak-https.json_ to the _spec_ folder. Added a spec that will test if a grant can be obtained directly by using https. I'm not sure if this is what you mean. I think you will need to change your test environment (WildFly) for this spec, because it requires a https listener.
I struggled a bit to get the existing specs working with Jasmine, because I don't have the Keycloak's realm (_nodejs-test_)/client (_nodejs-auth-utils_) configuration. For example, the tests only seem to succeed when access tokens expire very quickly (less than 2 seconds).
Maybe it's an idea to add an export of this configuration to this project as well?

abstractj commented:
@rockxwre @lholmquist correct me if I'm wrong, but I believe that with the latest changes from @lance. This PR was also covered, nope? See: https://github.com/keycloak/keycloak-nodejs-auth-utils/blob/master/lib/grant-manager.js#L146-L151

lholmquist commented:
@abstractj right,  looks like it

lance commented:
Actually, no my PR did not cover these changes.

rockxwre commented:
@abstractj I agree with @lance latest comment: these changes do not cover this PR. Still no https support for a few functions.

abstractj commented:
@rockxwre I'm sorry if I overlooked these changes. Could you please rebase? I started doing it, but the chances of myself doing catastrophic mess with your changes are high.

rockxwre commented:
@abstractj I see that @lance already has committed (f77482f4a384e4791ed064449b290b6cfda896c4) my changes in his own repository. Should I still rebase or will @lance take care of it?

lance commented:
I will submit a PR today

On Monday, May 9, 2016, Ramon Rockx <notifications@github.com> wrote:

> @abstractj <https://github.com/abstractj> I see that @lance
> <https://github.com/lance> already has committed (f77482f
> <https://github.com/keycloak/keycloak-nodejs-auth-utils/commit/f77482f4a384e4791ed064449b290b6cfda896c4>)
> my changes in his own repository. Should I still rebase or will @lance
> <https://github.com/lance> take care of it?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/keycloak/keycloak-nodejs-auth-utils/pull/2#issuecomment-217817327>
>


-- 
Lance Ball
http://lanceball.com

lance commented:
@abstractj I think you can close this PR now - all of the provided capabilities have been merged with https://github.com/keycloak/keycloak-nodejs-auth-utils/commit/f77482f4a384e4791ed064449b290b6cfda896c4

abstractj commented:
@lance dunno! @rockxwre thanks a million for your patience.
